api/v1alpha7/types.go

```diff
@@ -163,20 +163,83 @@ type RootVolume struct {
-	// Size is the size in GB of the volume.
-	Size int `json:"diskSize"`
-	// VolumeType is the volume type of the volume.
-	// If omitted, the default type will be used.
-	VolumeType string `json:"volumeType,omitempty"`
-	// AvailabilityZone is the volume availability zone to create the volume in.
+// blockDeviceStorage is the storage type of a block device to create and
+// contains additional storage options.
+// +kubebuilder:validation:XValidation:rule="self.type == 'Volume' || !has(self.volume)",message="volume is forbidden when type is not Volume"
+// +union
+//
+//nolint:godot
+type BlockDeviceStorage struct {
+	// type is the type of block device to create.
+	// This can be either "Volume" or "Local".
+	// +kubebuilder:validation:Enum="Volume";"Local"
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	Type BlockDeviceType `json:"type"`
+
+	// volume contains additional storage options for a volume block device.
+	// +optional
+	// +unionMember,optional
+	Volume *BlockDeviceVolume `json:"volume,omitempty"`
+}
+
+// blockDeviceVolume contains additional storage options for a volume block device.
+type BlockDeviceVolume struct {
+	// type is the Cinder volume type of the volume.
+	// If omitted, the default Cinder volume type that is configured in the OpenStack cloud
+	// will be used.
+	// The maximum length of a volume type name is 255 characters, as per the OpenStack limit.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=255
+	// +optional
+	Type string `json:"type,omitempty"`
+
+	// availabilityZone is the volume availability zone to create the volume in.
 	// If omitted, the availability zone of the server will be used.
+	// The availability zone must NOT contain spaces otherwise it will lead to volume that belongs
+	// to this availability zone register failure, see kubernetes/cloud-provider-openstack#1379 for
+	// further information.
+	// The maximum length of availability zone name is 63 as per labels limits.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:XValidation:rule="!self.contains(' ')",message="availabilityZone may not contain spaces"
+	// +optional
 	AvailabilityZone string `json:"availabilityZone,omitempty"`
 }
 
+// additionalBlockDevice is a block device to attach to the server.
+type AdditionalBlockDevice struct {
+	// name of the block device in the context of a machine.
+	// If the block device is a volume, the Cinder volume will be named
+	// as a combination of the machine name and this name.
+	// Also, this name will be used for tagging the block device.
+	// Information about the block device tag can be obtained from the OpenStack
+	// metadata API or the config drive.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// sizeGiB is the size of the block device in gibibytes (GiB).
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Required
+	SizeGiB int `json:"sizeGiB"`
```

Comment:
I'm not a massive fan of field names that contain the unit. Is it not sufficient for the description and docs to say the unit?

Reply:
If you prefer not to have the units in the name, then you should use a `resource.Quantity`, which would mean the field would look like `size: 128Gi` to the user.

This is as per current [Kube API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units).

I'm happy with either but @EmilienM preferred this over the `resource.Quantity` IIUC

Reply:
Not a strong opinion, but for me it seems preferable to have it part of the name.

Reply:
having `size: 128Gi` would be error prone IMO (we would certainly need some validation). I prefer to have the unit in the name. Unless someone has a strong opinion.